lib/src/package_analysis/lower_bound_constraint_analysis.dart

```diff
@@ -0,0 +1,378 @@
+    } else {
+      context.warning(
+          'Attempting to get a resolved unit for file $filePath resulted in an invalid result.');
+    }
+  }
+
+  return astVisitor.issues.values.whereNotNull().toList();
+}
+
+class _LowerBoundConstraintVisitor extends GeneralizingAstVisitor {
+  /// The [PackageAnalysisContext] corresponding to the package being analyzed.
+  final PackageAnalysisContext context;
+
+  /// The [Element] definition being analyzed.
+  late Element element;
+
+  /// The id of [element].
+  late int elementId;
+
+  // The name of the package which defines [element].
+  late String packageName;
+
+  // The summary of the package which defines [element].
+  late PackageShape dependencyShape;
+
+  /// The parent of [element].
+  late Element parentElement;
+
+  /// The [SourceSpan] corresponding to the invocation currently being analyzed.
+  late SourceSpan span;
+
+  /// The name of the symbol being analyzed.
+  late String symbolName;
+
+  /// The package URI for the currently analyzed unit, typically on the form
+  /// `package:package_name/library_name.dart`.
+  late Uri currentUri;
+
+  /// Maps from [Element.id] to either [LowerBoundConstraintIssue] if the lower
+  /// bound constraint on [LowerBoundConstraintIssue.dependencyPackageName] is
+  /// too low, making [Element] when the lowest allowed version
+  /// [LowerBoundConstraintIssue.lowestVersion] is used, or null otherwise.
+  final Map<int, LowerBoundConstraintIssue?> issues = {};
+
+  /// The summaries of each of the dependencies of the target package.
+  final Map<String, PackageShape> dependencySummaries;
+
+  _LowerBoundConstraintVisitor({
+    required this.context,
+    required this.dependencySummaries,
+  });
+
+  /// Attempt to identify the name of the defining library of [element],
+  /// populating [packageName] and [dependencyShape].
+  void identifyDependencyName() {
+    if (parentElement.library == null) {
+      throw AnalysisException(
+          'Could not determine library of parentElement.');
+    }
+    final tryPackageName =
+        packageFromLibraryUri(parentElement.library!.identifier);
+    // if the defining package isn't a HostedDependency of the target, then
+    // this symbol cannot be analyzed
+    if (tryPackageName == null ||
+        !context.targetDependencies.keys.contains(tryPackageName)) {
+      throw AnalysisException(
+          'The defining package is not a HostedDependency of the target package.');
+    }
+    packageName = tryPackageName;
+
+    if (!dependencySummaries.keys.contains(packageName)) {
+      context.warning('No summary for $packageName found.');
+      throw AnalysisException('No summary was found for the defining package.');
+    }
+    dependencyShape = dependencySummaries[packageName]!;
+  }
+
+  /// Populate the various properties of this visitor based on information from
+  /// this symbol, throwing an [AnalysisException] if the symbol does not need
+  /// to be/cannot be analyzed.
```

Comment:
We could also return some data class that holds the different properties, that might be less confusing.